NetKet/Output/json_output_writer.hpp

```diff
@@ -111,13 +111,14 @@ class JsonOutputWriter {
   // The first overload works for classes inheriting from AbstractMachine, the
   // second one for Eigen matrices.
   template <typename T>
-  void SaveState_Impl(std::ofstream& stream, const AbstractMachine<T>& state) {
-    state.Save(stream);
+  void SaveState_Impl(std::ofstream& stream,
+                      std::shared_ptr<AbstractMachine<T>> state) {
+    state->Save(stream);
   }
 
   template <typename T, int S1, int S2>
   void SaveState_Impl(std::ofstream& stream,
-                      const Eigen::Matrix<T, S1, S2>& state) {
+                      Eigen::Ref<Eigen::Matrix<T, S1, S2>> state) {
```

Comment:
I think this should be a `Ref<const Eigen::Matrix...>` since the function is not supposed to modify the matrix.